pkg/elasticetcd/doc.go

```diff
@@ -0,0 +1,24 @@
+// Package elasticetcd implements an autoscaling etcd cluster.
+// This package was inspired by, and implements the elastic scaling algorithm as described in, the github.com/purpleidea/mgmt/etcd package.
+//
+// The elastic scaling algorithm is as follows,
+// 	- The elasticetcd instance is started, and can be passed a list of existing etcd endpoints
+//  	- If no endpoints are given, assume you are the first instance up and start your embedded etcd server
+//	- Connect to the given endpoints (or your own etcd endpoints if you were the first one up)
+//	- Volunteer to be a server, and wait to be nominated as a server
+//		- If you are nominated, start your embedded etcd server and join the existing cluster
+//		- If your nomination is removed, stop your embedded etcd server
+//	- Begin a campaign to become the leader of the elastic cluster
+// 		- When elected as the leader, make nominations from the volunteer list, to keep the right number of servers.
+// 		- Watch for changes to the volunteer list, online servers and the ideal size, and make/remove nominations as required.
+//
+// Right now the server nominations are selected in a round-robin fashion, using the list of volunteers sorted by name.
+//
```

Comment:
ElasticEtcd stop wasn't being done in a safe way. The watchers were left running, while the client was stopped. 
The client nomination would be removed as the session ended when it was stopped.
There was a brief moment where the nomination watcher was triggered before it was canceled, leading to the `doNominations` handler being run. This would panic when it tried to use the now nil `ee.cli`. This panic would prevent the etcd server from cleanly transferring the leadership and stopping, leading to an inconsistent cluster.

This is now fixed, with the watchers being stopped first before stopping the client.